File: python/63.py
import unittest
from typing import List


class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        dp: List[List[int]] = [[0] * len(obstacleGrid[0])
                               for _ in range(len(obstacleGrid))]

        for i in range(len(obstacleGrid)):
            for j in range(len(obstacleGrid[0])):
                if obstacleGrid[i][j] == 1:
                    continue

                if i == 0 and j == 0:
                    dp[i][j] = 1

                if i - 1 >= 0 and i - 1 < len(obstacleGrid) and obstacleGrid[i - 1][j] != 1:
                    dp[i][j] += dp[i - 1][j]

                if j - 1 >= 0 and obstacleGrid[i][j - 1] != 1:
                    dp[i][j] += dp[i][j - 1]

        return dp[len(obstacleGrid) - 1][len(obstacleGrid[0]) - 1]


class TestSolution(unittest.TestCase):
    def test(self):
        solution = Solution()
        self.assertEqual(solution.uniquePathsWithObstacles(
            [[0, 0, 0], [0, 1, 0], [0, 0, 0]]), 2)

File: typescript/63.ts
function uniquePathsWithObstacles(obstacleGrid: number[][]): number {
    const dp: number[][] = new Array(obstacleGrid.length).fill([]).map(() => new Array(obstacleGrid[0].length).fill(0));
    for (let i = 0; i < obstacleGrid.length; i++) {
        for (let j = 0; j < obstacleGrid[0].length; j++) {
            if (obstacleGrid[i][j] === 1) {
                dp[i][j] = 0;
                continue;
            }
            if (i === 0 && j === 0) {
                dp[i][j] = 1;
                continue;
            }
            if (i === 0) {
                dp[i][j] = dp[i][j - 1];
                continue;
            }
            if (j === 0) {
                dp[i][j] = dp[i - 1][j];
                continue;
            }
            dp[i][j] = dp[i - 1][j] + dp[i][j - 1];
        }
    }
    return dp[obstacleGrid.length - 1][obstacleGrid[0].length - 1];
};

test('63', () => {
    expect(uniquePathsWithObstacles([[0, 0, 0], [0, 1, 0], [0, 0, 0]])).toBe(2);
})